Add vector embeddings for search queries

jsdanielh opened the pull request:
Add vector embedding for search queries.

This closes #11.

jsdanielh committed:
Add database schema and migration for vector support

jsdanielh committed:
Create embedding generation utilities and OpenAI inregration

jsdanielh committed:
Implement hybrid search functionality in the search endpoint

Use a bash script for generating embeddings since node is not
available in the database container image.

jsdanielh committed:
Add environment configuration and validation

onmax commented on database/schema.ts:
```diff
@@ -38,4 +53,5 @@ export const locationCategories = pgTable('location_categories', {
 ])
 
+export type Location = typeof locations.$inferSelect
 export type LocationCategory = typeof locationCategories.$inferSelect
```
This was my mistake. This kind of types are defined in `~/shared/types`. defining the types here seems to no give us anything.

onmax commented on database/startup.sh:
```diff
@@ -18,5 +18,9 @@ echo "Running migrations on startup..."
 /docker-entrypoint-initdb.d/03-seed.sh || true
 
+# Run the vector embeddings generation
+echo "Running vector embeddings generation on startup..."
```
This morning the whole docker/supabase thing was refactored completely. I think this might create a mergin issue. I will fix it in this PR :)

onmax commented on server/api/search.get.ts:
```diff
@@ -156,60 +211,42 @@ export default defineEventHandler(async (event): Promise<LocationResponse[]> =>
+      consola.warn('OpenAI authentication failed, falling back to keyword-only search')
+    }
+    else if (error instanceof Error && error.message.includes('vector')) {
+      consola.warn('Vector database operations failed, falling back to keyword-only search')
```
i would remove this completely. Just use

`throw createError(error)`

`createError` is a function from h3.

if we needed it we can always add more details later, but let;s keep the code clean i would say

onmax commented on server/utils/config-validation.ts:
```diff
@@ -0,0 +1,202 @@
+    if (config.similarityThreshold < 0.3) {
+      warnings.push('VECTOR_SEARCH_SIMILARITY_THRESHOLD is very low (<0.3). This may result in many irrelevant vector search results.')
+    }
+  }
```
This kind of check adds complexity. we can always improve the `safeRuntimeConfig` to add these checks.

onmax commented on server/plugins/config-validation.ts:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Nitro plugin to validate configuration during server startup
+ */
+export default defineNitroPlugin(async (_nitroApp) => {
```
since the validateVectorSearchConfig is unnecessary, this whole plugin is also unnecessary imo

onmax commented on server/utils/config-validation.ts:
```diff
@@ -0,0 +1,202 @@
+  /**
+   * Validate all required environment variables and configuration
+   */
+  static validateConfiguration(): void {
```
not needed this whole file

onmax commented on server/utils/embedding.ts:
```diff
@@ -0,0 +1,253 @@
+/**
+ * Service for generating embeddings using OpenAI API
+ */
+export class EmbeddingService {
```
I think this is way too much code.

we can use something like this https://ai-sdk.dev/docs/ai-sdk-core/embeddings#embedding-a-single-value

Then setup the OPENAI_API_KEY in .env and it should be more than enough.

onmax replied:
config such as maxRetries: 3,
  baseDelay: 1000, // 1 second
  maxDelay: 30000, // 30 seconds
  batchSize: 100, // OpenAI batch limit

not necessaary in m opinion. the ai sdk provides that for us

onmax replied:
also the "batch" optimization is also not ncessary.

I would suggest to run a embedding manually for the categories and put the result in the `database/seed` The scrit can be run with tsx and ai sdk. We should run it only once.

in the future if we update the categories, then we run it again.

onmax reviewed (changes requested):
would you like to have a call?

onmax committed:
Remove Docker-based embedding generation setup

onmax committed:
Apply PR review feedback

- Remove Location type from schema (use shared/types)
- Simplify error handling in search endpoint
- Remove config validation (use safeRuntimeConfig)
- Remove location embeddings (only category embeddings)
- Revert to named catalog structure in workspace
- Remove vector search dependencies and config

onmax committed:
Revert schema changes

onmax committed:
Implement hybrid search with FTS and embeddings

- Add full-text search on location name and address
- Add category similarity search using embeddings
- Cache query embeddings in NuxtHub KV (30 day TTL)
- Combine text and category results with deduplication
- Use AI SDK for embedding generation
- Support user-selected category filters

onmax committed:
Add autocomplete and embedding precompute endpoints

- Add /api/search/autocomplete for fast full-text search
- Add /api/search/embed for background embedding precompute
- Remove TTL from embedding cache (permanent storage)
- Min query length: 2 characters

onmax committed:
Implement autocomplete UI and simplify selects

- Add form-based autocomplete with search icon
- Show user's query as first result
- Display matching locations below with map pin icons
- Precompute embeddings while typing (background)
- Support UUID-based location fetching
- Extract locationSelect helper to reduce duplication
- Use semantic search when submitting user's query
- Use direct UUID fetch when selecting location

onmax committed:
Refactor UI and add match highlighting

- Format template to one line per element
- Use @submit.prevent for form submission
- Remove icons from autocomplete locations
- Add PostgreSQL ts_headline for match highlighting
- Style <mark> tags as bold (no background)
- Move search.get.ts to search/index.get.ts

onmax committed:
fix: update lockfile and dependency versions

onmax committed:
docs: improve comments to explain why, add data flow diagram

onmax committed:
fix: skip runtime config validation during build

onmax committed:
fix: use DATABASE_URL and fix import paths

onmax committed:
chore: remove old supabase folder

onmax committed:
fix: allow empty strings in runtime config validation

onmax committed:
refactor: split index.vue logic into composables

onmax committed:
refactor: create locations/[uuid] endpoint, move opening hours to utils

onmax committed:
feat: add i18n for opening hours messages

onmax committed:
refactor: include messageKey in opening hours status

onmax committed:
refactor: move location fetch to component, add location enrichment utils

onmax committed:
refactor: simplify location endpoint with single query

onmax committed:
refactor: simplify createError calls, use validated params, add i18n translations

onmax committed:
refactor: optimize queries and simplify abstractions

onmax committed:
refactor: simplify search, remove created_at

onmax committed:
docs: update docs and simplify filters to booleans
